docs/runtime-model.md

```diff
@@ -0,0 +1,356 @@
+struct Count {
+  remaining: usize,
+}
+
+impl Future for Count {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        while self.count > 0 {
+            self.count -= 1;
+
+            // Yield every 10 iterations
+            if self.count % 10 == 0 {
+                task::current().notify();
+                return Ok(Async::NotReady);
+            }
+        }
+
+        Ok(Async::Ready(()))
+    }
+}
+```
+
+[contract]: https://docs.rs/futures/0.1.23/futures/future/trait.Future.html#tymethod.poll
+
+### Executors
+
+Executors are responsible for driving many tasks to completion. A task is
+spawned onto an executor, at which point the executor calls its `poll` function
+when needed. The executor hooks into the task system to receive resource
+readiness notifications.
+
+By decoupling the task system with the executor implementation, the specific
+execution and scheduling logic can be left to the executor implementation. Tokio
+provides two executor implementations, each with unique characteristics:
+[`current_thread`] and [`thread_pool`].
+
+When a task is first spawned onto the executor, the executor wraps it w/
+[`Spawn`][Spawn]. This binds the task logic with the task state (this is mostly
+required for legacy reasons). When the executor picks a task for execution, it
+calls [`Spawn::poll_future_notify`][poll_future_notify]. This function ensures
+that the task context is set to the thread-local variable such that
+[`task::current`][current] is able to read it. When calling
+[`poll_future_notify`][poll_future_notify], the executor also passes in a notify
+handle and an identifier. These arguments are included in the task handle
+returned by [`task::current`][current] and is how the task is linked to the
+executor.
+
+The notify handle is an implementation of [`Notify`][`Notify`] and the identifier
+is a value that the executor uses to look up the current task. When
+[`Task::notify`][notify] is called, the [`notify`][Notify::notify] function on
+the notify handle is called with the supplied identifier. The implementation of
+this function is responsible for performing the scheduling logic.
+
+One strategy for implementing an executor is to store each task in a `Box` and
+to use a linked list to track tasks that are scheduled for execution. When
+[`Notify::notify`][Notify::notify] is called, then the task associated with the
+identifier is pushed at the end of the `scheduled` linked list. When the
+executor runs, it pops from the front of the linked list and executes the task
+as described above.
+
+Note that this section does not describe how the executor is run. The details of
+this are left to the executor implementation. One option is for the executor to
+spawn one or more threads and dedicate these threads to draining the `scheduled`
+linked list. Another is to provide a `MyExecutor::run` function that blocks the
+current thread and drains the `scheduled` linked list.
```

Comment:
Link to some examples of different implementations?

Reply:
Other implementation strategies are mostly theoretical at this point. I am not aware of any to link to.